File: scala/ch02/ex2.38.scala
def foldRight[A, B] (op: (A, B) => B, initial: B, sequence: List[A]): B =
  sequence match {
    case Nil => initial
    case _ => op (sequence.head, foldRight(op, initial, sequence.tail))
  }

def foldLeft[A, B] (op: (A, B) => B, initial: B, sequence: List[A]): B = {
  def iter(result: B, rest:List[A]): B =
    rest match {
      case Nil => result
      case _ => iter(op (rest.head, result), rest.tail)
    }
  iter(initial, sequence)
}

import org.scalatest._ 
import org.scalatest.matchers._

object foldLeftRightSpec extends Spec with ShouldMatchers {
  describe ("foldLeft") {
    it ("should return a result of a left fold") {
      foldLeft((x:Double, y:Double) => y/x, 1.0, List(1.0, 2.0, 3.0)) should equal (1.0/6.0)
      foldLeft((x:Int, l:List[_]) => List(l,x), Nil, List(1, 2, 3)) should equal (List(List(List(Nil, 1), 2), 3))
    }
  }
  describe ("foldRight") {
    it ("should return a result of a right fold") {
      foldRight((x:Double, y:Double) => x/y, 1.0, List(1.0, 2.0, 3.0)) should equal (3.0/2.0)
      foldRight((x:Int, l:List[_]) => List(x,l), Nil, List(1, 2, 3)) should equal (List(1, List(2, List(3, Nil))))
    }
  }
}
foldLeftRightSpec execute

File: clojure/ch02/ex2.38.clj
(ns sicp.ch02 (:use clojure.contrib.test-is))

(defn fold-right [op initial sequence]
  (if (empty? sequence)
      initial
      (op (first sequence)
          (fold-right op initial (rest sequence)))))

(defn fold-left [op initial sequence]
  (loop [result initial seq sequence]
    (if (empty? seq)
        result
        (recur (op result (first seq))
              (rest seq)))))



(deftest test-fold-left-right
  (is (= (fold-right / 1 (list 1 2 3)) (/ 3 2)))

  (is (= (fold-left / 1 (list 1 2 3)) (/ 1 6)))

  (is (= (fold-right list (list) (list 1 2 3)) (list 1 (list 2 (list 3 (list))))))

  (is (= (fold-left list (list) (list 1 2 3)) (list (list (list (list) 1) 2) 3))))

; For fold-right and fold-left to return the same results for any sequence, "op"
; must be commutative, e.g., + and *:

(deftest test-fold-left-right-commutative-for-mult-add
  (is (= (fold-right + 0 (list 1 2 3 4)) 10))
  (is (= (fold-left  + 0 (list 1 2 3 4)) 10))
  (is (= (fold-right * 1 (list 1 2 3 4)) 24))
  (is (= (fold-left  * 1 (list 1 2 3 4)) 24)))

(run-tests)
